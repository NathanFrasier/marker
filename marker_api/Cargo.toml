[package]
name = "marker_api"
version = "0.1.0"
edition = "2021"
license = "MIT OR Apache-2.0"
readme = "./README.md"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
visibility = "0.0.1"

[features]
# Some items should only be used by the driver implementing the functionality,
# this feature enables the export of these items. Note that this interface is
# unstable. All usage of the driver api can change between releases.
driver-api = []
