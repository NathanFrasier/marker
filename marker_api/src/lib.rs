#![doc = include_str!("../README.md")]
#![warn(clippy::pedantic)]
#![warn(clippy::exhaustive_enums)]
#![allow(clippy::module_name_repetitions)]
#![allow(clippy::must_use_candidate)]
#![allow(clippy::trivially_copy_pass_by_ref)]
#![allow(clippy::unused_self)] // `self` is needed to change the behavior later
#![allow(clippy::missing_panics_doc)] // Temporary allow for `todo!`s
#![allow(clippy::new_without_default)] // Not very helpful as `new` is almost always cfged
#![cfg_attr(not(feature = "driver-api"), allow(dead_code))]

pub static MARKER_API_VERSION: &str = env!("CARGO_PKG_VERSION");

pub mod ast;
pub mod context;
pub mod diagnostic;
pub mod interface;
pub mod lint;

#[doc(hidden)]
pub mod ffi;

pub use context::AstContext;
pub use interface::{LintPassInfo, LintPassInfoBuilder};

/// A [`LintPass`] visits every node like a `Visitor`. The difference is that a
/// [`LintPass`] provides some additional information about the implemented lints.
/// The adapter will walk through the entire AST once and give each node to the
/// registered [`LintPass`]es.
pub trait LintPass {
    fn info(&self) -> LintPassInfo;

    fn check_item<'ast>(&mut self, _cx: &'ast AstContext<'ast>, _item: ast::item::ItemKind<'ast>) {}
    fn check_field<'ast>(&mut self, _cx: &'ast AstContext<'ast>, _field: &'ast ast::item::Field<'ast>) {}
    fn check_variant<'ast>(&mut self, _cx: &'ast AstContext<'ast>, _variant: &'ast ast::item::EnumVariant<'ast>) {}
    fn check_body<'ast>(&mut self, _cx: &'ast AstContext<'ast>, _body: &'ast ast::item::Body<'ast>) {}
    fn check_stmt<'ast>(&mut self, _cx: &'ast AstContext<'ast>, _stmt: ast::stmt::StmtKind<'ast>) {}
    fn check_expr<'ast>(&mut self, _cx: &'ast AstContext<'ast>, _expr: ast::expr::ExprKind<'ast>) {}
}
