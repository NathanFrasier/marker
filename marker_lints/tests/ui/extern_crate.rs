extern crate marker_api;

use marker_api::{ast::expr::ExprKind, context::AstContext};

marker_api::declare_lint!(
    DUMMY,
    Warn,
    "dummy",
);

pub fn check_expr<'ast>(cx: &AstContext<'ast>, expr: ExprKind<'ast>) {
    cx.emit_lint(DUMMY, expr.id(), "X <-- starting with upper case", expr.span(), |_| {});
}

fn main() {}
