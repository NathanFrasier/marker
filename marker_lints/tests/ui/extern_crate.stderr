warning: X <-- starting with upper case
  --> $DIR/extern_crate.rs:11:70
   |
11 |   pub fn check_expr<'ast>(cx: &AstContext<'ast>, expr: ExprKind<'ast>) {
   |  ______________________________________________________________________^
12 | |     cx.emit_lint(DUMMY, expr.id(), "X <-- starting with upper case", expr.span(), |_| {});
13 | | }
   | |_^
   |
   = note: `#[warn(marker::diag_msg_capital_start)]` on by default

warning: X <-- starting with upper case
  --> $DIR/extern_crate.rs:15:11
   |
15 | fn main() {}
   |           ^^

warning: 2 warnings emitted

